### ISSUE TITLE
"Did you mean" spelling suggestions can offer the same thing being complained about

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#25564]
**TypeScript Version:** 2.92

**Search Terms:** did you mean spelling suggestion same

**Code**

```ts
type NonVoid = Object | object;

type Something<T> = { test: string } & (T extends NonVoid ? {
    requiredWhenNonVoid: T
} : {
    requiredWhenNonVoid?: undefined
    });

function testFunc2<A extends NonVoid>(a: A): Something<A> {
    return {
        test: 'test',
        requiredWhenNonVoid: a
//      ~~~~~~~~~~~~~~~~~~~
    };
}
```

**Expected behavior:** Probably no error, or at least not this error suggestion

**Actual behavior:** 

```
clipboard.ts:12         requiredWhenNonVoid: a
                        ~~~~~~~~~~~~~~~~~~~~~~
  TS2322: Type '{ test: string; requiredWhenNonVoid: A; }' is not assignable to type 'Something<A>'.
    Object literal may only specify known properties, but 'requiredWhenNonVoid' does not exist in type 'Something<A>'. Did you mean to write 'requiredWhenNonVoid'?
```

There's no spelling error here -- the `requiredWhenNonVoid` being cited is the same `requiredWhenNonVoid` being suggested 🤡 

**Playground Link:** [Link](http://www.typescriptlang.org/play/#src=interface%20Foo%20%7B%20hello%3A%20string%20%7D%3B%0A%0Afunction%20bar1(...foos%3A%20Foo%5B%5D)%20%7B%0A%20%20%20%20bar2(...foos)%3B%0A%7D%0A%0Afunction%20bar2(foo1%3A%20Foo)%3B%0Afunction%20bar2(foo1%3A%20Foo%2C%20foo2%3A%20Foo)%3B%0Afunction%20bar2(...foos%3A%20Foo%5B%5D)%20%7B%0A%20%20%20%20return%3B%0A%7D%0A)

**Related Issues:** Nope

**Comment by jcalz** [comment]
> an assert that the result is not equal to the missing name

So then a real misspelling suggests the correct spelling, which is then an error?  Sounds fun:

```ts
return {
  test: 'test',
  requiredWhenNotVoid: a
//~~~~~~~~~~~~~~~~~~~ Error:
//Object literal may only specify known properties,
//but 'requiredWhenNotVoid' does not exist in 
//type 'Something<A>'. Did you mean to write 'requiredWhenNonVoid'?
}
```
Okay...

```ts
return {
  test: 'test',
  requiredWhenNonVoid: a  // fixed
//~~~~~~~~~~~~~~~~~~~ Error:
//Object literal may only specify known properties,
//but 'requiredWhenNonVoid' does not exist in 
//type 'Something<A>'. Ha ha, fooled you! 😝
}
```
I hate you, compiler!  🏃‍♂️ 😭

**Comment by sandersn** [comment]
I figured out how to fix the excess property error at least: make excess property checks understand conditional types:

```ts
/** (in isKnownProperty) */
            else if (targetType.flags & TypeFlags.Conditional) {
                return isKnownProperty((targetType as ConditionalType).root.trueType, name, isComparingJsxAttributes) ||
                    isKnownProperty((targetType as ConditionalType).root.falseType, name, isComparingJsxAttributes);
            }
```

This exposes the more inscrutable bug that @RyanCavanaugh and I encountered, which is that `{ requiredWhenNonVoid: a }` is not assignable to `Something<A>`. This is probably a lack of assignability rules for conditional types, but I haven't worked out whether it's inherent to the definition of conditionals or just an oversight &mdash; new types typically start with few assignability rules.

**Comment by tjenkinson** [comment]
Thanks for fixing the message so quickly! Any idea why the example still doesn't compile though? It should be valid right?

This came from my initial question here: https://stackoverflow.com/q/51274107/1048589

**Comment by sandersn** [comment]
Basically, there's no assignability rule that directly relates object types to conditional types, and because we're inside `testFunc2`, we can't (I think) collapse the conditional type down to either its true branch or false branch. So we're looking at the following assignability check, as expressed in my own made-up notation:

```
{ arg: A } ==> A extends object ? { arg: A } : { arg: undefined } 
    where A extends object
```

The assignment the other way works because of a simplification rule that applies:

```
> A extends object ? { arg: A } : { arg: undefined } ==> { arg: A }
    where A extends object
> object extends object ? { arg: A} : {arg: undefined } ==> { arg: A }
> { arg: A } ==> { arg: A }
#t
```

So I know why the example doesn't compile. I don't know if the reason is valid: Should the simplification rule also apply when the target is a conditional type, or just when the source is a conditional type? I'm not sure. @ahejlsberg have you thought about this assignability case?

**Comment by tjenkinson** [comment]
@ahejlsberg any ideas? I'm wondering if I should create a separate issue for tracking this? This causes an issue for us currently and we have to revert to some '<any>' trickery and it would be great to not have to do this.

**Comment by tjenkinson** [comment]
@ahejlsberg thanks I understand a bit better what is happening now.

I think there is actually already a ticket for this: https://github.com/Microsoft/TypeScript/issues/24929
